### PR TITLE
add lwip net stack

### PR DESCRIPTION
**Pull request by Centaurus99**
~Working In Progress.~
✅ Ready for review

- [x] pbuf copy optimization
- [x] lwip parameter optimization

### CODE REVIEW

File: scripts/make/qemu.mk [hunk 1]
```diff
@@ -30,9 +30,19 @@ qemu_args-$(FS) += \
...
+ifeq ($(NETDEV), user)
+  qemu_args-$(NET) += \
+    -netdev user,id=net0,hostfwd=tcp::5555-:5555,hostfwd=udp::5555-:5555
+else ifeq ($(NETDEV), tap)
+  qemu_args-$(NET) += \
+    -netdev tap,id=net0,ifname=qemu-tap0,script=no,downscript=no
+endif
+
+ifeq ($(MODE), debug)
+  qemu_args-$(NET) += \
```
Comment (equation314) [inline scripts/make/qemu.mk]: `qemu_args-debug` is used when `MODE=debug`

File: apps/c/httpclient/expect_info.out [hunk 2]
```diff
@@ -18,6 +18,7 @@ Initialize device drivers...
 registered a new Net device at .\+: "virtio-net"
 Initialize network subsystem...
   use NIC 0: "virtio-net"
+Using smoltcp
```
Comment (equation314) [inline apps/c/httpclient/expect_info.out]: Add a two-space indent:

```
  net stack: smoltcp
```

File: apps/c/httpclient/expect_info_lwip.out [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ .bss (READ | WRITE | RESERVED)
+ free memory (READ | WRITE | FREE)
+Initialize global memory allocator...
+Initialize kernel page table...
+Initialize platform devices...
+Initialize device drivers...
+registered a new Net device at .\+: "virtio-net"
+Initialize network subsystem...
+  use NIC 0: "virtio-net"
+Using lwip
```
Comment (equation314) [inline apps/c/httpclient/expect_info_lwip.out]: Same as above

File: scripts/make/qemu.mk (hunk 1 above)
Comment (Centaurus99) [reply scripts/make/qemu.mk]: `qemu_args-debug` is not used when `MODE=debug`. It is used for gdb debugging when run with `make debug`.

Here I intend to capture packets when `MODE=debug`, regardless of whether gdb is being used or not.

Comment (equation314) [reply scripts/make/qemu.mk]: Maybe a better argument, e.g., `NET_DUMP=y`?

File: modules/axnet/src/smoltcp_impl/mod.rs [hunk 4]
```diff
@@ -296,6 +296,7 @@ fn snoop_tcp_packet(buf: &[u8]) -> Result<(), smoltcp::wire::Error> {
 }
 
 pub(crate) fn init(mut net_dev: AxNetDevice) {
+    info!("  net stack: smoltcp");
```
Comment (equation314) [inline modules/axnet/src/smoltcp_impl/mod.rs]: Show this info in "axnet/lib.rs"

File: crates/lwip_rust/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,10 @@
+[package]
+name = "lwip_rust"
+version = "0.1.0"
+edition = "2021"
+
```
Comment (equation314) [inline crates/lwip_rust/Cargo.toml]: Add more keys, see other crates.

File: Makefile [hunk 6]
```diff
@@ -44,20 +45,27 @@ else
...
 # Binutils
-ifeq ($(APP_LANG), c)
-  CROSS_COMPILE ?= $(ARCH)-linux-musl-
-  CC := $(CROSS_COMPILE)gcc
-  AR := $(CROSS_COMPILE)ar
-  RANLIB := $(CROSS_COMPILE)ranlib
-  LD := rust-lld -flavor gnu
-endif
+CROSS_COMPILE ?= $(ARCH)-linux-musl-
+CC := $(CROSS_COMPILE)gcc
```
Comment (equation314) [inline Makefile:67]: All use `?=`

Comment (Centaurus99) [reply Makefile:67]: For some reason, using `?=` will cause CC to be set to the string `cc`.

Comment (equation314) [reply Makefile:67]: OK, we can set `CROSS_COMPILE=xxxx-` to change the default CC.

File: crates/lwip_rust/build.rs [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    // Not build with clippy or doc
+    if target_os == "none" && clippy_args.is_err() {
+        compile_lwip();
+    }
+    generate_lwip_bindings();
+}
+
+fn generate_lwip_bindings() {
+    println!("cargo:rustc-link-lib=lwip");
+    println!("cargo:include=depend/lwip/src/include/");
```
Comment (equation314) [inline crates/lwip_rust/build.rs]: `cargo:rerun-if-changed=custom` and `cargo:rerun-if-changed=depend`

File: crates/lwip_rust/build.rs [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        .file("depend/lwip/src/core/ipv6/ip6_addr.c")
+        .file("depend/lwip/src/core/ipv6/ip6_frag.c")
+        .file("depend/lwip/src/core/ipv6/mld6.c")
+        .file("depend/lwip/src/core/ipv6/nd6.c")
+        .file("depend/lwip/src/netif/ethernet.c")
+        // .file("depend/lwip/src/apps/lwiperf/lwiperf.c")
+        .file("custom/sys_arch.c");
+
+    base_config
+        .warnings(false)
```
Comment (equation314) [inline crates/lwip_rust/build.rs]: enable warnings

File: Makefile
Comment (equation314) [inline Makefile:1]:
```makefile
ifeq ($(ARCH), x86_64)
  ACCEL ?= y
  PLATFORM ?= pc-x86
  TARGET := x86_64-unknown-none
  TARGET_CFLAGS := -mabi=lp64d
  BUS := pci
else ifeq ($(ARCH), riscv64)
  ACCEL ?= n
  PLATFORM ?= qemu-virt-riscv
  TARGET := riscv64gc-unknown-none-elf
  TARGET_CFLAGS := -mno-sse
else ifeq ($(ARCH), aarch64)
  ACCEL ?= n
  PLATFORM ?= qemu-virt-aarch64
  TARGET := aarch64-unknown-none-softfloat
  TARGET_CFLAGS := -mgeneral-regs-only
else
  $(error "ARCH" must be one of "x86_64", "riscv64", or "aarch64")
endif
```

Add these after set `CC`:

```makefile
export TARGET_CC=$(CC)	# for building lwip_rust
export TARGET_CFLAGS
```

File: ulib/c_libax/build.mk [hunk 9]
```diff
@@ -26,15 +30,9 @@ ifneq ($(wildcard $(in_feat)),)	# check if features.txt contains "fp_simd"
 endif
 
 ifeq ($(ARCH), riscv64)
-  CFLAGS += -march=rv64gc -mabi=lp64d -mcmodel=medany
```
Comment (equation314) [inline ulib/c_libax/build.mk]: Restore these changes

File: scripts/make/cargo.mk [hunk 10]
```diff
@@ -53,8 +54,12 @@ endif
 
 rustc_flags := -Clink-args="-T$(LD_SCRIPT) -no-pie"
 
+ifeq ($(ARCH), x86_64)
+  rustc_flags += -Clink-args="--no-relax"
+endif
+
 define cargo_build
-  cargo rustc $(build_args) $(1) -- $(rustc_flags)
```
Comment (equation314) [inline scripts/make/cargo.mk:58]: Restore it.

**Comment by chyyuu** [comment]
@Centaurus99 dev merged newest main, then  the axnet build failed for 'Err(AxError::Again)'.  Please fix this little bug. Thanks!

**Comment by Centaurus99** [comment]
> @Centaurus99 dev merged newest main, then the axnet build failed for 'Err(AxError::Again)'. Please fix this little bug. Thanks!

Fixed. But note that lwip_impl does not yet support iperf app.

**Comment by chyyuu** [comment]
@Centaurus99 Try fix further, and support iperf app.

**Pull request merged**